### PR TITLE
add ability to check array of roles

### PR DESCRIPTION
**Pull request by fiqrisr**
Add ability to check array of roles like #554

### CODE REVIEW

File: src/check-roles.ts [hunk 1]
```diff
@@ -2,12 +2,22 @@ import { ForbiddenException, UnauthorizedException } from '@nestjs/common';
...
 ) {
+  let parsedRole = role;
+
+  // parse role to array if role is array in comma seperated string (like in header)
+  if (typeof role === 'string') {
+    parsedRole =
+      role.replace(/\s/g, '').split(',').length > 1
+        ? (role.replace(/\s/g, '').trim().split(',') as any as R[])
+        : (role as R);
+  }
```
Comment (iamolegga) [inline src/check-roles.ts:15]: this should be done out of scope of the current lib. it's the logic of specific unmarshalling, in your case it should already be done in the passed callback and return final result

Comment (iamolegga) [inline src/check-roles.ts:15]: this should be done out of scope of the current lib. it's the logic of specific unmarshalling, in your case it should already be done in the passed callback and return final result

File: src/check-roles.ts [hunk 2]
```diff
@@ -27,13 +37,27 @@ export function checkRoles<R>(
...
+      !(requiredRoles as [R, ...R[]]).some((requiredRole) =>
+        (parsedRole as R[]).includes(requiredRole),
+      )
+    ) {
+      throw new ForbiddenException();
+    }
+
+    // if role is single role
+  } else {
+    if (!(requiredRoles as [R, ...R[]]).includes(parsedRole as R)) {
```
Comment (iamolegga) [inline src/check-roles.ts:58]: please return comment for this line

File: README.md [hunk 3]
```diff
@@ -60,7 +60,9 @@ function getRole(context: ExecutionContext) {
   if (!session) {
     return;
   }
-  return (session as { role?: Role }).role;
+  
+  // you can use single role or array of roles
+  return (session as { role?: Role | Role[] }).role;
```
Comment (iamolegga) [inline README.md:65]: please add comment that if **at least one** role of array is allowed, guard will pass request

File: src/check-roles.ts (hunk 1 above)
Comment (fiqrisr) [reply src/check-roles.ts:15]: Okay but the test will fail because header doesn't support array. Should the test changed?

Comment (iamolegga) [reply src/check-roles.ts:15]: you should not change tests, just change [this line](https://github.com/iamolegga/nestjs-roles/blob/master/__tests__/index.spec.ts#L15)

Comment (fiqrisr) [reply src/check-roles.ts]: okay done

File: src/check-roles.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@ import { ForbiddenException, UnauthorizedException } from '@nestjs/common';
 
 export function checkRoles<R>(
   requiredRoles: [R, ...R[]] | [boolean],
-  role: R | undefined,
+  role: R | R[] | string | undefined,
```
Comment (iamolegga) [inline src/check-roles.ts:5]: ~string~

Comment (fiqrisr) [reply src/check-roles.ts:5]: sorry forgot to delete that

Comment (fiqrisr) [reply src/check-roles.ts]: done

File: __tests__/index.spec.ts [hunk 5]
```diff
@@ -42,6 +43,11 @@ for (const platform of platforms) {
         expect(fn).toHaveBeenCalledTimes(1);
       });
 
+      it('should pass users with array of roles', async () => {
+        await createTest(controller, Guard, platform, [Role.ADMIN]);
```
Comment (iamolegga) [inline __tests__/index.spec.ts:47]: let's add roles to array here

File: __tests__/index.spec.ts [hunk 6]
```diff
@@ -67,6 +73,11 @@ for (const platform of platforms) {
         expect(fn).toHaveBeenCalledTimes(0);
       });
 
+      it('should not pass users with array of roles', async () => {
+        await createTest(controller, Guard, platform, [Role.ADMIN]);
```
Comment (iamolegga) [inline __tests__/index.spec.ts:77]: let's add roles to array here

File: __tests__/index.spec.ts [hunk 7]
```diff
@@ -92,6 +103,11 @@ for (const platform of platforms) {
         expect(fn).toHaveBeenCalledTimes(1);
       });
 
+      it('should pass users with array of roles', async () => {
+        await createTest(controller, Guard, platform, [Role.ADMIN]);
```
Comment (iamolegga) [inline __tests__/index.spec.ts:107]: let's add roles to array here

File: __tests__/index.spec.ts [hunk 8]
```diff
@@ -117,11 +133,24 @@ for (const platform of platforms) {
         expect(fn).toHaveBeenCalledTimes(1);
       });
 
+      it('should pass users with allowed roles in array', async () => {
+        await createTest(controller, Guard, platform, [Role.ADMIN]);
```
Comment (iamolegga) [inline __tests__/index.spec.ts:137]: let's add roles to array here

Comment (fiqrisr) [reply __tests__/index.spec.ts:137]: when I add more than one role here, the test failed becasue when I console.log header, the role property is in comma seperated string, not in array (with square brackets)

Comment (iamolegga) [reply __tests__/index.spec.ts:137]: As I've said before you need to change [this line](https://github.com/iamolegga/nestjs-roles/blob/master/__tests__/index.spec.ts#L15) to check if this header's value is a string with comma and then parse it to array

Comment (iamolegga) [reply __tests__/index.spec.ts:137]: then it will be correct tests for the new feature

Comment (fiqrisr) [reply __tests__/index.spec.ts:137]: Ahhh I see

**Comment by iamolegga** [comment]
Ah, there is an issue with the CI, will update soon

**Comment by fiqrisr** [comment]
> Ah, there is an issue with the CI, will update soon

Why it takes so long?

**Comment by iamolegga** [comment]
> Why it takes so long?

@fiqrisr because I'm doing this for free in my free time. You can always publish and use your own package

**Comment by fiqrisr** [comment]
> > Why it takes so long?
> 
> @fiqrisr because I'm doing this for free in my free time. You can always publish and use your own package

I'm sorry, I just hope it will get merge soon because this is my first pull request to open source

**Comment by iamolegga** [comment]
@fiqrisr no worry, will fix this soon

**Comment by iamolegga** [comment]
@fiqrisr please rebase on top of master to get fixed ci for PR

**Comment by fiqrisr** [comment]
> @fiqrisr please rebase on top of master to get fixed ci for PR

done

**Comment by iamolegga** [comment]
released in `0.5.1`

**Pull request merged**